core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArrayAggregationAfterDistinct.java

```diff
@@ -0,0 +1,176 @@
+        return aggregationNode.getAggregations().values().stream()
+                .anyMatch(aggregation -> aggregation.getResolvedFunction().getSignature().getName().equals(ArrayAggregationFunction.NAME));
+    }
+
+    private static boolean hasArrayDistinct(Metadata metadata, ProjectNode projectNode)
+    {
+        return projectNode.getAssignments().getExpressions().stream()
+                .anyMatch(assignment -> assignment instanceof FunctionCall &&
+                        metadata.decodeFunction(((FunctionCall) assignment).getName()).getSignature().getName().equals(ArrayDistinctFunction.NAME));
+    }
+
+    @Override
+    public Result apply(ProjectNode projectNode, Captures captures, Context context)
+    {
+        AggregationNode aggregationNode = captures.get(CHILD);
+
+        Assignments.Builder assignmentsBuilder = Assignments.builder();
+        ImmutableMap.Builder<Symbol, Aggregation> aggregationsBuilder = ImmutableMap.builder();
+
+        List<Symbol> matchedAggregations = new ArrayList<>();
+        List<Symbol> matchedAssignments = new ArrayList<>();
+
+        Set<Symbol> arrayAggSymbols = aggregationNode.getAggregations().entrySet().stream()
+                .filter(entry -> entry.getValue().getResolvedFunction().getSignature().getName().equals(ArrayAggregationFunction.NAME))
+                .map(Map.Entry::getKey)
+                .collect(toImmutableSet());
+
+        Map<Symbol, Expression> arrayDistinctCalls = projectNode.getAssignments().entrySet().stream()
+                .filter(entry -> entry.getValue() instanceof FunctionCall)
+                .filter(entry -> {
+                    FunctionCall functionCall = (FunctionCall) entry.getValue();
+                    return functionCall.getArguments().size() == 1 &&
+                            arrayAggSymbols.contains(Symbol.from(getOnlyElement(functionCall.getArguments())));
+                })
+                // We only decode function calls that have an array aggregation as only parameter (because decoding is expensive)
+                .filter(entry -> {
+                    FunctionCall functionCall = (FunctionCall) entry.getValue();
+                    ResolvedFunction resolvedFunction = metadata.decodeFunction(functionCall.getName());
+                    return resolvedFunction.getSignature().getName().equals(ArrayDistinctFunction.NAME);
+                })
+                .collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
+
+        Set<Symbol> referencedOutsideArrayDistinctAggregateSymbols = projectNode.getAssignments().entrySet().stream()
+                .filter(entry -> !arrayDistinctCalls.containsKey(entry.getKey()))
+                .flatMap(entry -> SymbolsExtractor.extractUnique(entry.getValue()).stream())
+                .filter(arrayAggSymbols::contains)
+                .collect(toImmutableSet());
+
+        Map<Symbol, Aggregation> arrayAggregationsWithArrayDistinct = aggregationNode.getAggregations().entrySet().stream()
```

Comment:
This also captures any aggregations which are not referenced in the ProjectNode. They shouldn't be rewritten, even if it does not matter as they are not used.